torchrl/envs/utils.py

```diff
@@ -219,44 +219,63 @@ def _set_single_key(source, dest, key, clone=False):
     if isinstance(key, str):
         key = (key,)
     for k in key:
-        val = source.get(k)
-        if is_tensor_collection(val):
-            new_val = dest.get(k, None)
-            if new_val is None:
-                new_val = val.empty()
-                # dest.set(k, new_val)
-                dest._set_str(k, new_val, inplace=False, validated=True)
-            source = val
-            dest = new_val
+        shape = source.get_item_shape(k)
+        if -1 not in shape:
+            val = source.get(k)
+            if is_tensor_collection(val):
+                new_val = dest.get(k, None)
+                if new_val is None:
+                    new_val = val.empty()
+                    # dest.set(k, new_val)
+                    dest._set_str(k, new_val, inplace=False, validated=True)
+                source = val
+                dest = new_val
+            else:
+                if clone:
+                    val = val.clone()
+                # dest.set(k, val)
+                dest._set_str(k, val, inplace=False, validated=True)
         else:
-            if clone:
-                val = val.clone()
-            # dest.set(k, val)
-            dest._set_str(k, val, inplace=False, validated=True)
+            # this is a het key
+            for s_td, d_td in zip(source.tensordicts, dest.tensordicts):
+                _set_single_key(s_td, d_td, k, clone)
+            break
 
 
 def _set(source, dest, key, total_key, excluded):
     total_key = total_key + (key,)
     non_empty = False
     if unravel_key(total_key) not in excluded:
-        val = source.get(key)
-        if is_tensor_collection(val):
-            new_val = dest.get(key, None)
-            if new_val is None:
-                new_val = val.empty()
+        shape = source.get_item_shape(key)
+        if -1 not in shape:
```

Comment:
what's the overhead here?
Since this will likely be very frequent I'm a bit worried by increasing the control flow for every key

Reply:
For a regular td, this op means that we'll be accessing the tensor twice through get (once to get its shape, a second time to get it). Accessing it once should suffice.
Not I understand that we can't use `get` if the tensor is heterogeneous... A try/except may not be the right solution either since it'll be slow for heterogeneous envs. However regular envs won't be impacted

Reply:
I am not sure what I can do here
- in the past i was using try blocks, but you made me move away from them
- an isinstance wont work since the key might be het even if the container is a regular td

so I can move back to try except if you prefer

Reply:
lemme switch to try blocks

ideally tho this problem could be solved once and for all in tensordict, with an efficient way to understand if a key is het

Reply:
yep that's what I'm thinking too
If we have tensorstack we can even leave the block with try / except and it won't break (I guess)

Reply:
Just make sure you comment the try / excpet saying that it's temporary and for efficiency purpose